mkiterian/bucket-list-app-api
# #149365897 Add pagination to get bucketlist and items data

### What does this PR do?
- Implement pagination so that users can specify a limit and/or page of bucketlists or items to receive from a get request
- Add a search feature so that a user can search for a bucketlist by name
- Fix bug where user could modify other users bucketlists
#### Description of Task to be completed?
##### Pagination
- Logged in users can receive a response with data containing their bucketlists via get requests through the endpoint:
    - _GET /api/v1/bucketlists_
- This endpoint returns all the bucketlists belonging to the logged in user
- By adding a **'limit'** parameter to the url, the user can specify exactly how many bucketlists or items to receive. e.g
    - _GET /api/v1/bucketlists?limit=5_
    - _This will return the first 5 bucketlists(ordered by id in ascending order) belonging to the user_
- The user can also specify a **'page'** parameter which refers to a set of bucketlists by limit
    - _GET /api/v1/bucketlists?limit=5&page=2_
    - _This will return the second set of 5 bucketlists for the user by ascending id order i.e bucketlists 6th to 10th_
- The default limit is set to 10, for when a GET url request contains **'page'** parameter but no **'limit'**_
    - _GET /api/v1/bucketlists?page=3_
    - _This will return the third set of 10 bucketlists i.e (bucketlists 21 - 30 by order)_
- The above also applies to GET requests on the endoint:
    - _/api/v1/bucketlists/id/items_
##### Search
- By adding a parameter **'q'** a user can search for a bucketlist by name by entering a string/substring of the bucketlist name
    -  _GET /api/v1/bucketlists?q=fun will return a response with bucketlists whose names contain 'fun' e.g Fun, Fundamentals, Funny, Morefun_
##### Bug Fix
- User could be able to update other user's bucketlist. Fixed by limiting user to access only the user's bucketlists and items
#### How should this be manually tested?
- The API endpoints can be tested using Postman or curl via the Terminal
_Site root is http://127.0.0.1:5000 where not specified_
- **Testing pagination** e.g for a get request on the /api/v1/bucketlists endpoint:
    - _Start the server locally by running; python run.py in the terminal_
    - _Login with valid credentials to receive an authentication token_
    - _In a new tab set the request type to GET_
    - _Enter the endpoint url in the Postman url field as:                     http://127.0.0.1:5000/api/v1/bucketlists?limit=3_
    - _Input the token received from login into the header with the key as Authorization. The token should be prefixed with "JWT "_
    - _Click on the GET button to send the request_
    - _You should receive a response with three bucketlists (if they exist)_
- **Testing search** 
    - _Create multiple bcuketlists via _POST /api/v1/bucketlists_. Name 2 with a similar subtring e.g 'ave'. Send a GET request: /api/v1/bucketlists?q=ave, should return in the response the two bucketlists whose names contain the matching substring_ 
#### What are the relevant pivotal tracker stories?
#149365897
#149861850
#149365775
